Namespaced Wasm Imports so they don't conflict across modules, or reserved LLVM IR

torch2424 opened the pull request:
closes #1643
closes #1559

Kind of relates to #1647
Doesn't fix, but puts on the right path I think, for relates to #1648

So this PR was opened from a discussion in the tinygo slack: https://gophers.slack.com/archives/CDJD3SUP6/p1614201204223100 . Involving a conflict with wasm imports with the name of `write`. Which conflicts with the LLVM IR `@write`, and requires decoupling the LLVM IR symbol, and the `wasm-import-name`.

I tested this pr with the following code: https://github.com/torch2424/tinygo-compiler-playground (Which also contains the output LLVM IR of this current PR) , and ran the tests: `go test -target=wasm` and `make wasmtest`. 

### Screenshots of tests and things

![Screenshot from 2021-02-24 18-20-02](https://user-images.githubusercontent.com/1448289/109099144-76510700-76d7-11eb-8aee-3c568c71b758.png)
![Screenshot from 2021-02-24 18-15-57](https://user-images.githubusercontent.com/1448289/109099146-76e99d80-76d7-11eb-8904-26703ba9a264.png)
![Screenshot from 2021-02-24 18-14-04](https://user-images.githubusercontent.com/1448289/109099147-77823400-76d7-11eb-9371-e23d0071b06e.png)

torch2424 committed:
Started logging out some stuff to figure out what's going on in external
functions

torch2424 committed:
Started taking a look at the Symbol.go code for the wasi fix

torch2424 committed:
Got the function name to correctly change in the symbols

torch2424 committed:
Namespaced all of the wasm import functions

torch2424 committed:
Fixed issues with no module imports

torch2424 commented:
Oops, looks like my fork got a little weird, Let me force push to this branch based on my current branch :smile: 

EDIT: Ah! The actual issue was that I tried to merge into `release`, and not `dev` :upside_down_face: Fixing that now!

EDIT2: This is now for real good to go! :smile: :tada:

torch2424 committed:
Fixed the PR so it's actually good to go

deadprogram commented:
Hello @torch2424 thanks for this contribution.

Looks like there are some problems with the CI build: https://app.circleci.com/pipelines/github/tinygo-org/tinygo/4409/workflows/09f56573-cf9f-4d30-af51-fc15dfc75753/jobs/24355/parallel-runs/0/steps/0-121?invite=true

It appears that you need to run `go fmt` on the test output.

Separately, please make sure you rebase off of the most current `dev` branch to pickup some CI fixes from yesterday/

Thank you!

torch2424 committed:
Ran go fmt, and removed a newline

torch2424 commented:
Hello @deadprogram Thanks for the tips! :smile: 

> Looks like there are some problems with the CI build: https://app.circleci.com/pipelines/github/tinygo-org/tinygo/4409/workflows/09f56573-cf9f-4d30-af51-fc15dfc75753/jobs/24355/parallel-runs/0/steps/0-121?invite=true

> It appears that you need to run go fmt on the test output.

So I tried running `go fmt` and `gofmt -w` but it just gives the same error unfortunately :( 

![Screenshot from 2021-02-25 11-13-29](https://user-images.githubusercontent.com/1448289/109204745-937ae980-775a-11eb-8a4e-99e03395de74.png)

Though, I am a bit confused why this would be happening? As I don't know why the test output is incorrectly formatted for the fix that I made? :thinking: 

> Separately, please make sure you rebase off of the most current dev branch to pickup some CI fixes from yesterday/

I made sure to do that! Looking at my branch, I am now rebased on top of: https://github.com/torch2424/tinygo/commit/e76729c6f39772542f0e4f4f85f5a41f655fa505 . Which seems to match up with what is at the HEAD of `tinygo:dev` :smile: :tada:

torch2424 committed:
Ran the fmt from the makefile

torch2424 commented:
> So I tried running `go fmt` and `gofmt -w` but it just gives the same error unfortunately :( 

So I did some digging around the makefile, and I just had to run the fmt in the makefile, `make fmt`.  :smile: 

---

Things are starting to pass in the CI, so I think this is good to go @deadprogram ! :smile: :tada:

aykevl commented on compiler/symbol.go:
```diff
@@ -151,8 +153,13 @@ func (c *compilerContext) getFunction(fn *ssa.Function) llvm.Value {
 			wasmImportModuleAttr := c.ctx.CreateStringAttribute("wasm-import-module", info.module)
 			llvmFn.AddFunctionAttr(wasmImportModuleAttr)
+
+			wasmImportNameAttr := c.ctx.CreateStringAttribute("wasm-import-name", info.importName)
+			llvmFn.AddFunctionAttr(wasmImportNameAttr)
```
Perhaps this needs an extra check that `info.importName` is not unset (an empty string)?

aykevl commented on compiler/symbol.go:
```diff
@@ -207,7 +214,16 @@ func (info *functionInfo) parsePragmas(f *ssa.Function) {
+				if info.module == "" {
+					info.linkName = info.importName
+				} else {
+					info.linkName = fmt.Sprintf("tinygo_wasm_import_%s_%s", info.module, info.importName)
```
You are here introducing an order dependency between `//go:wasm-module` and `//export`: using `//export` before `//go:wasm-module` would change the behavior which is probably not expected.
Also, you're setting `info.importName` for all systems, not just WebAssembly? I guess LLVM will ignore it but it doesn't seem clean to me.

Instead of this if/else, I would move this if/else right after the for loop (putting `parts[1]` in a temporary variable). Something like this:

```go
if info.module != "" {
    // WebAssembly import
    info.importName = importName
} else {
    info.linkName = importName
}
```

Note, in particular, that this _doesn't_ change the link name. I think that's fine (and perhaps better): a TinyGo function name is unlikely to conflict with a name from another programming language with names such as `main.someExportedName` or `github.com/user/pkg/subpkg.someFunction`.

torch2424 committed:
Started making the suggested fixes, but got a little stuck

torch2424 commented:
Thanks for the review @aykevl ! :smile: 

So I addressed the first comment! 

and then for the second suggestion I tried moving it after the for loop, but then I get runtime panics (I added a screenshot), and I couldn't figure out why that was happening. :thinking: But! I was going to start heading out for the day, so I just pushed up where I'm at. But no worries, I'm sure I'll figure out what is going on. :smile: 

![Screenshot from 2021-02-25 19-15-18](https://user-images.githubusercontent.com/1448289/109250290-47ea2f00-779e-11eb-9503-6b9d622b40b5.png)

Once everything is good to go I'll add a mention :smile: :+1:

torch2424 committed:
Made requested changes

torch2424 committed:
Ran the formatter

torch2424 committed:
Removed a bad comment

torch2424 commented:
@deadprogram @aykevl Alright! So I fixed the bug I had yesterday! :smile: :tada: 

I saw the CI was passing, but then found an uneeded comment, removed it, and then re-ran `make fmt`. And my example is still working! :tada: 

![Screenshot from 2021-02-26 13-04-07](https://user-images.githubusercontent.com/1448289/109354771-28511600-7833-11eb-9975-16d82921b8e6.png)


So I think this is good to go! :smile:

deadprogram commented:
This is some great work @torch2424 thank you for all of the updates to feedback.

Would you prefer that I squash the commits before merging, or would you like to do that yourself?

aykevl commented on compiler/symbol.go:
```diff
@@ -191,6 +200,10 @@ func (info *functionInfo) parsePragmas(f *ssa.Function) {
 	if decl, ok := f.Syntax().(*ast.FuncDecl); ok && decl.Doc != nil {
+
+		// Our importName for a wasm module (if we are compiling to wasm), or llvm link name
+		var importName string = ""
```
Just a nitpick: this can be declared slightly shorter as follows:

```go
importName := ""
```

or like this:

```go
var importName string
```

aykevl reviewed (approved):
Looks good to me!

torch2424 committed:
Made requested changes

torch2424 commented:
Ah! My apologies, been handling some family stuff as of lately! :smile: 

@deadprogram 

> Would you prefer that I squash the commits before merging, or would you like to do that yourself?

First, thank you for the kind words! :smile: :tada: Please feel free to squash this, I'd very much appreciate it :smile: 

---

@aykevl

Thank you for the review. Made the suggested change, this should be good to go! :smile: :tada:

deadprogram commented:
Thank you so very much for all the effort on this contribution. Now squash/merging.

torch2424 commented:
@deadprogram @aykevl Yay! Thank you very much for the review, and squashing / merging! :smile: :tada:
